### PR TITLE
Pablo/invoices live

### PR DESCRIPTION
**Pull request by Pabloj20**
Customer_name and reference data are not the correct ones. I have just rendered some random fields from the invoice schema to see if it works

### CODE REVIEW

**Comment by Pabloj20** [comment]
Captura de pantalla de 2021-12-22 12-43-58

**Comment by Pabloj20** [comment]
#24

**Review by peillis — CHANGES_REQUESTED** [review]
change data structures

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+       invisible_buttons: invisible_buttons,
+       checked: update_checked
+     )}
+  end
+
+  defp someone_checked?(checked) do
+    checked
+    |> Enum.empty?()
+  end
+end
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:40]: Think about the state. There can be "no possible" situations with your actual variables. For example there could be checked invoices and the buttons be invisible.
Remember to design a state where there can't be "impossible" states.

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  def mount(_params, _session, socket) do
+    {:ok,
+     socket
+     |> assign(:invoices, Invoices.list())
+     |> assign(:invisible_buttons, true)
+     |> assign(:checked, [])}
+  end
+
+  def handle_event("show_buttons", %{"id" => id, "value" => "on"}, socket) do
+    update_checked = socket.assigns.checked ++ [id]
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:14]:
```suggestion
    update_checked = [id | socket.assigns.checked]
```

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 3]
```diff
@@ -0,0 +1,40 @@
+defmodule SiwappWeb.InvoicesLive.Index do
+  use SiwappWeb, :live_view
+  alias Siwapp.Invoices
+
+  def mount(_params, _session, socket) do
+    {:ok,
+     socket
+     |> assign(:invoices, Invoices.list())
+     |> assign(:invisible_buttons, true)
+     |> assign(:checked, [])}
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:10]: MapSet would be better than a list. MapSet guarantees there is no item repeated

File: lib/siwapp_web/live/live_helpers.ex [hunk 4]
```diff
@@ -12,4 +12,18 @@ defmodule SiwappWeb.LiveHelpers do
     modal_opts = [id: :modal, return_to: path, component: component, opts: opts]
     live_component(SiwappWeb.ModalComponent, modal_opts)
   end
+
+  def checked?(checked, id) do
+    string_id = Integer.to_string(id)
+
+    checked_id =
+      checked
+      |> Enum.find(fn n -> n == string_id end)
```
Comment (peillis) [inline lib/siwapp_web/live/live_helpers.ex:21]: Enum.member? is better for this

File: lib/siwapp_web/live/live_helpers.ex [hunk 5]
```diff
@@ -12,4 +12,18 @@ defmodule SiwappWeb.LiveHelpers do
...
+
+    checked_id != nil
+  end
+
+  def pending?(due_date) do
```
Comment (peillis) [inline lib/siwapp_web/live/live_helpers.ex:26]: pending is more complicated than this. If the invoice is paid it's not pending.

**Review by peillis — CHANGES_REQUESTED** [review]

**Review by peillis — CHANGES_REQUESTED** [review]
fix

**Review by peillis — CHANGES_REQUESTED** [review]
think! damn it!

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+         socket,
+         checked: update_checked
+       )}
+    end
+  end
+
+  defp add_all_ids(invoices) do
+    add_zero =
+      MapSet.new()
+      |> MapSet.put("0")
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:61]: You can write this as MapSet.new(["0"])

**Review by peillis — CHANGES_REQUESTED** [review]
rf the index.ex

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    {:noreply,
+     assign(
+       socket,
+       checked: checked
+     )}
+  end
+
+  defp update_checked(%{"id" => "0", "value" => "on"}, socket) do
+    add_zero =
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:23]: bad name

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      MapSet.new()
+      |> MapSet.put("0")
+
+    socket.assigns.invoices
+    |> Enum.reduce(add_zero, fn invoice, mapset ->
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:28]: wrongly use of reduce

**Review by peillis — CHANGES_REQUESTED** [review]

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  def mount(_params, _session, socket) do
+    {:ok,
+     socket
+     |> assign(:invoices, Invoices.list())
+     |> assign(:checked, MapSet.new())}
+  end
+
+  def handle_event("click_checkbox", data, socket) do
+    checked = update_checked(data, socket)
+    IO.inspect(checked)
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:14]: I see you don't check your PRs before me. Please, ALWAYS do that, and don't tell me I forgot to delete that. 
What you forgot is to see the changes you want to merge.

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  alias Siwapp.Invoices
+
+  def mount(_params, _session, socket) do
+    {:ok,
+     socket
+     |> assign(:invoices, Invoices.list())
+     |> assign(:checked, MapSet.new())}
+  end
+
+  def handle_event("click_checkbox", data, socket) do
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:12]: as a convention what you call here data is usually called params. Please rename that.

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    {:noreply,
+     assign(
+       socket,
+       checked: checked
+     )}
+  end
+
+  defp update_checked(%{"id" => "0", "value" => "on"}, socket) do
+    socket.assigns.invoices
+    |> MapSet.new(fn invoice -> invoice.id end)
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:25]:
```suggestion
    |> MapSet.new(& &1.id)
```

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+  def handle_event("click_checkbox", data, socket) do
+    checked = update_checked(data, socket)
+    IO.inspect(checked)
+
+    {:noreply,
+     assign(
+       socket,
+       checked: checked
+     )}
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:20]: all this tuple in a single line would be more readable

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  defp update_checked(%{"id" => id, "value" => "on"}, socket) do
+    MapSet.put(socket.assigns.checked, String.to_integer(id))
+  end
+
+  defp update_checked(%{"id" => id}, socket) do
+    socket.assigns.checked
+    |> MapSet.delete(String.to_integer(id))
+    |> MapSet.delete(0)
+  end
+end
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:42]: Good, bit by bit is getting much more simple. You should compare this code with your initial proposal.

File: lib/siwapp_web/live/live_helpers.ex [hunk 14]
```diff
@@ -12,4 +12,24 @@ defmodule SiwappWeb.LiveHelpers do
     modal_opts = [id: :modal, return_to: path, component: component, opts: opts]
     live_component(SiwappWeb.ModalComponent, modal_opts)
   end
+
```
Comment (peillis) [inline lib/siwapp_web/live/live_helpers.ex:15]: write specs for these fns below

File: lib/siwapp_web/live/live_helpers.ex [hunk 15]
```diff
@@ -12,4 +12,24 @@ defmodule SiwappWeb.LiveHelpers do
...
+  def checked?(checked, id) do
+    Enum.member?(checked, id)
+  end
+
+  def status(paid, due_date) do
```
Comment (peillis) [inline lib/siwapp_web/live/live_helpers.ex:20]: This status fn should go to the Invoices context. It's not just a helper for liveview, it's something more internal of every invoice.

File: lib/siwapp_web/live/live_helpers.ex [hunk 16]
```diff
@@ -12,4 +12,24 @@ defmodule SiwappWeb.LiveHelpers do
...
+    cond do
+      paid -> :paid
+      pending?(due_date) -> :pending
+      true -> :past_due
+    end
+  end
+
+  def invisible?(checked) do
```
Comment (peillis) [inline lib/siwapp_web/live/live_helpers.ex:28]: instead of writing a fn which does just 1 thing, use the fn below and use it directly in the template

**Review by peillis — CHANGES_REQUESTED** [review]

File: lib/siwapp/invoices.ex [hunk 17]
```diff
@@ -65,4 +65,17 @@ defmodule Siwapp.Invoices do
   def get_by!(key, value) do
     Repo.get_by!(Invoice, %{key => value})
   end
+
+  def status_per_id() do
+    list()
+    |> Map.new(fn invoice -> {invoice.id, status(invoice.paid, invoice.due_date)} end)
+  end
```
Comment (peillis) [inline lib/siwapp/invoices.ex:72]: Esta fn no tiene mucho sentido. Lo ideal sería tener una tipo `status(invoice)`

File: lib/siwapp_web/live/invoices_live/index.ex [hunk 18]
```diff
@@ -0,0 +1,38 @@
+defmodule SiwappWeb.InvoicesLive.Index do
+  use SiwappWeb, :live_view
+  alias Siwapp.Invoices
+
+  def mount(_params, _session, socket) do
+    {:ok,
+     socket
+     |> assign(:invoices, Invoices.list())
+     |> assign(:checked, MapSet.new())
+     |> assign(:status, Invoices.status_per_id())}
```
Comment (peillis) [inline lib/siwapp_web/live/invoices_live/index.ex:10]: No hace falta esta variable en el socket, ya tienes las invoices.

**Review by peillis — CHANGES_REQUESTED** [review]
Además corrige los errores que salen del `mix credo --strict`
Gracias

File: lib/siwapp/invoices.ex [hunk 19]
```diff
@@ -65,4 +65,12 @@ defmodule Siwapp.Invoices do
   def get_by!(key, value) do
     Repo.get_by!(Invoice, %{key => value})
   end
+
+  def status(invoice) do
```
Comment (peillis) [inline lib/siwapp/invoices.ex:69]: Escribe @specs please

**Review by peillis — APPROVED** [review]

**Pull request merged**